cmd/agent/app/session.go

```diff
@@ -1,230 +1,148 @@
-}
-
-func (session *Session) State() string {
-	session.mutex.RLock()
-	defer session.mutex.RUnlock()
-
-	return session.state
-}
-
-func (session *Session) ActiveConnections() *orderedmap.OrderedMap[string, *Reference[connection.Connection]] {
-	session.mutex.RLock()
-	defer session.mutex.RUnlock()
-
-	// activeConnections := orderedmap.New[string, *Reference[connection.Connection]]()
-
-	// for pair := session.connections.Oldest(); pair != nil; pair = pair.Next() {
-	// 	connection := pair.Value.Object
-	// 	if connection.ExitStatus() == restapi.ExitStatusUnknown {
-	// 		activeConnections.Set(pair.Key, pair.Value)
-	// 	}
-	// }
-
-	// return activeConnections
-
-	return session.connections
-}
-
-func NewSession(id string, juicePath string, version string, gpus *gpu.SelectedGpuSet, eventListener EventListener, persistent bool) *Session {
-	session := Session{
-		id:            id,
+func newSession(ctx context.Context, id string, version string, persistent bool, juicePath string, gpus *gpu.SelectedGpuSet, eventListener EventListener) *Session {
+	return &Session{
+		Id:            id,
+		Version:       version,
+		Persistent:    persistent,
 		juicePath:     juicePath,
-		version:       version,
-		state:         restapi.SessionActive,
 		gpus:          gpus,
-		persistant:    persistent,
-		connections:   orderedmap.New[string, *Reference[connection.Connection]](),
+		done:          make(chan struct{}),
+		connections:   utilities.NewConcurrentMap[string, *Connection](),
+		taskManager:   task.NewTaskManager(ctx),
 		eventListener: eventListener,
 	}
-
-	return &session
 }
 
 func (session *Session) Session() restapi.Session {
-	session.mutex.RLock()
-	defer session.mutex.RUnlock()
-
 	connections := make([]restapi.Connection, 0, session.connections.Len())
-	for pair := session.connections.Oldest(); pair != nil; pair = pair.Next() {
-		connections = append(connections, pair.Value.Object.Connection())
-	}
+	session.connections.Foreach(func(key string, value *Connection) bool {
+		connections = append(connections, restapi.Connection{
+			ConnectionData: value.ConnectionData,
+		})
+		return true
+	})
 
 	return restapi.Session{
-		Id:          session.id,
-		State:       session.state,
-		Version:     session.version,
+		Id:          session.Id,
+		State:       restapi.SessionActive,
+		Version:     session.Version,
 		Gpus:        session.gpus.GetGpus(),
 		Connections: connections,
-		Persistent:  session.persistant,
+		Persistent:  session.Persistent,
 	}
 }
 
-func (session *Session) Close() error {
-	errs := []error{}
-	// verify that there are no connections
-	if session.ActiveConnections().Len() > 0 {
-		logger.Errorf("closing session has active connections, %s", session.Id())
-		errs = append(errs, errors.New("closing session has active connections"))
-	}
-	session.mutex.Lock()
+func (session *Session) Run(group task.Group) error {
+	session.taskManager.GoFn(fmt.Sprintf("session %s close", session.Id), func(g task.Group) error {
```

Comment:
So each session will have it's own goroutine that will handling closing the session?

What happens if we call try to access a session while its closing? eg:

```
	session, found := agent.sessions.Get(sessionId)
```
We now have a reference to the session, but that session can be destroyed on another thread right?

Reply:
Good catch. Looking at this now.

Reply:
Resolved hopefully

Reply:
So in this restapi call
```
	session, err := agent.getSession(id)
	if err != nil {
		err = errors.Join(err, pkgnet.RespondWithString(w, http.StatusInternalServerError, err.Error()))
		logger.Error(err)
		return
	}

	err = pkgnet.Respond(w, http.StatusOK, session.Session())
	```
	
	I'm not sure I understand what this line does:
	```
	agent.taskManager.Go(fmt.Sprintf("session %s", sessionId), session)
	```
	Does this allow us to wait for all pending restapi calls to finish before closing the session?

Reply:
session when returned will be a valid pointer. The go garbage collector won't clean it up. The Session() function checks to see if it has been closed and returns the appropriate data.

Reply:
As for taskManager.Go, this is a wrapper function around a go routine in pkg/task. It associates a WaitGroup with all Go routines so when taskManager.Wait is called, it will wait until all GoRoutines exit before returning.